packages/devextreme/js/__internal/ui/chat/messagelist.ts

```diff
@@ -169,6 +173,89 @@ class MessageList extends Widget<Properties> {
     });
   }
 
+  _shouldAddDayHeader(
+    timestamp: undefined | string | number | Date,
+    currentMessageGroupItems: Message[] = [],
+  ): boolean {
+    const { showDayHeaders } = this.option();
+
+    if (!showDayHeaders) {
+      return false;
+    }
+
+    const deserializedDate = dateSerialization.deserializeDate(timestamp);
+
+    if (!isDate(deserializedDate) || isNaN(deserializedDate.getTime())) {
+      return false;
+    }
+
+    const lastDayHeaderDate = this._getLastDayHeaderDate(currentMessageGroupItems);
+
+    return !dateUtils.sameDate(lastDayHeaderDate, deserializedDate);
+  }
+
+  _getLastDayHeaderDate(currentMessageGroupItems: Message[] = []): Date | null {
+    let lastDayHeaderDate = null;
+    const messageGroups = this._messageGroups ?? [];
+
+    let index = currentMessageGroupItems.length - 1;
+
+    while (index >= 0 && lastDayHeaderDate === null) {
+      const messageTimestamp = currentMessageGroupItems[index].timestamp;
+
+      if (messageTimestamp) {
+        lastDayHeaderDate = dateSerialization.deserializeDate(messageTimestamp);
+      }
+      index -= 1;
+    }
+
+    index = messageGroups.length - 1;
+
+    while (index >= 0 && lastDayHeaderDate === null) {
+      const { items } = messageGroups[index].option();
+
+      const messageGroupTimestampList: (Date | string | number)[] = [];
+      items.forEach((item: Message) => {
+        if (item.timestamp) {
+          messageGroupTimestampList.push(item.timestamp);
+        }
+      });
+
+      if (messageGroupTimestampList.length > 0) {
+        lastDayHeaderDate = dateSerialization.deserializeDate(messageGroupTimestampList.at(-1));
+      }
+
+      index -= 1;
+    }
+
+    return lastDayHeaderDate;
+  }
```

Comment:
So, now we have 35-row function and 2 iteration through items instead of single cache variable using in the single place?
For what? It's much more difficult to maintain for me

Reply:
Yep, if we can't to write clear implementation it's better return single cache

Reply:
Okay, i see it was a request by Zhenya.
Let's then make this code more short and clear. Why should we process last message group separately from other ones?

```javascript
    const messageGroups = this._messageGroups ?? [];
    const messageGroupCount = messageGroupIndex.length;

    for (let messageGroupIndex = messageGroupCount - 1; messageGroupIndex >= 0; --messageGroupIndex) {
        const messageGroupItems = messageGroups[messageGroupIndex].option('items');
        const messagesCount = messageGroupItems.length;
        for (let messageIndex = messagesCount - 1; messageIndex >= 0; --messageIndex) {
            const messageTimestamp = messageGroupItems[messageIndex].timestamp;
            if (messageTimestamp) {
                const lastDayHeaderDate = dateSerialization.deserializeDate(messageTimestamp);
                return lastDayHeaderDate;
            }
        }
    }
```

Reply:
after some discussion, previous solution was reverted